scifight/utils.py

```diff
@@ -0,0 +1,39 @@
+import re
+
+
+def shorten_text(text, maxchars, ellipsis='...'):
+    """ Returns first few words from the `text`, but no more than `maxchars`, and appends
+        an `ellipsis`. This function works only on plain text (no HTML!), removes any leading
+        and trailing spaces, and squashes any number of consequent whitespaces (including
+        newlines) into a single space.
+
+        :param text: Input string, possibly quite long.
+        :param maxchars: Maximum length or result string.
+        :param ellipsis: Trailing sign to indicate implied continuation.
+        :return: Shortened substring with squashed whitespaces.
+    """
+    if not text or not isinstance(text, str) or maxchars <= 0:
+        return ''
+    if not ellipsis or not isinstance(ellipsis, str):
+        ellipsis = ''
+
+    if len(text) > maxchars:
+        counter = i = 0
+        for char in text:
+            if char not in {'\n', '\t', ' '}:
+                counter += 1
+            i += 1
+            if counter >= maxchars:
+                break
+        text = text[:i]
+
+    text = re.sub(r'\s+', ' ', text)
+    text = re.sub(r'^\s+', '', text)
+    text = re.sub(r'$\s+', '', text)[:maxchars+1]
```

Comment:
The cut is needed. First, we select all the characters except whitespace, but must return no more than the maxchars.